### PR TITLE
Move tests from doctest to Catch2

### PR DESCRIPTION
**Pull request by Gulin7**
Fixed: #297 

Tasks:

- [x] Modify dependency requirements to add Catch2:
   - [x] VCPKG: https://github.com/man-group/sparrow/blob/main/vcpkg.json
   - [x] Conda: https://github.com/man-group/sparrow/blob/main/environment-dev.yml
   - [x] Conan: https://github.com/man-group/sparrow/blob/main/conanfile.py
- [x] Modify CMake to use Catch2 library:
- [x] Modify CMake and create a new target which use Catch2: https://github.com/man-group/sparrow/blob/main/test/CMakeLists.txt
- [x] Modify run_tests and run_tests_with_junit_report targets to also run the new test target
- [x] Port a single unit test file: https://github.com/man-group/sparrow/blob/main/test/test_algorithm.cpp
- [x] Modify the CI github action files to upload both test report files: https://github.com/man-group/sparrow/blob/main/.github/workflows/linux.yml#L98

### CODE REVIEW

**Comment by Gulin7** [comment]
@Alex-PLACET I had problems with the set-up and I couldn't test locally, but I'll gladly test if you could give me a hand :)

**Comment by Gulin7** [comment]
There is something wrong with how I've set things up (it's my first time working directly with CMakes and vcpkg and so on).
When I build, I get this error: "test_algorithm.obj : error LNK2019: unresolved external symbol ..." multiple times.
image

**Comment by Alex-PLACET** [comment]
@Gulin7 According to [the documentation](https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md), you need to link the Catch2::Catch2WithMain cmake target and not Catch2::Catch2

**Comment by Gulin7** [comment]
@Alex-PLACET I've tried Catch2::Catch2WithMain and I get the same error as before

**Comment by Alex-PLACET** [comment]
[sparrow-17-13-20.patch](https://github.com/user-attachments/files/17937206/sparrow-17-13-20.patch)
With this patch, it compiles

**Comment by Gulin7** [comment]
Thank you so much, it worked! :) I've ran both the tests with Catch2 and the ones with Doctest and they were successful. I've pushed the changes as well.

**Comment by Alex-PLACET** [comment]
To fix the issue with MSVC compilation: 
[sparrow-09-22-45.patch](https://github.com/user-attachments/files/17944943/sparrow-09-22-45.patch)
Next thing is the issue with the CI: https://github.com/man-group/sparrow/actions/runs/12055549842/job/33616129801?pr=301
Check the annotation at the top.
And I guess it will be ready after the fixes

File: environment-dev.yml [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
+    - doctest
+    - catch2
+    # P0355R7 (Extending chrono to Calendars and Time Zones) has not been entirely implemented in libc++ yet.
+    # See: https://libcxx.llvm.org/Status/Cxx20.html#note-p0355
+    # For now, we use HowardHinnant/date as a replacement if we are compiling with libc++.
+    # TODO: remove this once libc++ has full support for P0355R7.
+    - howardhinnant_date
+    # Documentation
+    - doxygen
+    - graphviz
```
Comment (Alex-PLACET) [inline environment-dev.yml:18]: I'm not sure what is the issue but can you revert the indent be sure that you have blank line ?
https://github.com/man-group/sparrow/actions/runs/12066345620/job/33646992095?pr=301

**Comment by Alex-PLACET** [comment]
To fix the compilation on Windows with VCPKG, you have to force the build type to release:
next to https://github.com/man-group/sparrow/blob/main/.github/workflows/windows.yml#L128
`echo "VCPKG_BUILD_TYPE='Release'" >> $GITHUB_ENV`

**Comment by Alex-PLACET** [comment]
On exotic architectures, we get an old version of catch2 (version 2.13):
https://github.com/man-group/sparrow/actions/runs/12066345591/job/33646983709?pr=301#step:5:137
There is no cmake target Catch2::Catch2WithMain available with this version
In that case we need `#define CATCH_CONFIG_MAIN`
https://github.com/catchorg/Catch2/blob/v2.x/docs/tutorial.md#scaling-up
You have to create a main.cpp file which will be used only when we get Catch2 version < 3

**Comment by Gulin7** [comment]
Isn't there a way to enforce Catch2 >= 3.0.0?

**Comment by Alex-PLACET** [comment]
I will try to update exotic architectures images to ubuntu 22.04

**Comment by Alex-PLACET** [comment]
Hmmm, finally forcing VCPKG is not the clean way.
Conda don't provide dependencies in debug.
We should pass a define to CMAKE to force the use of the RuntimeLibrary in Debug or Release: https://cmake.org/cmake/help/latest/prop_tgt/MSVC_RUNTIME_LIBRARY.html when we use Conda

**Comment by Gulin7** [comment]
Would adding these inside if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
target_compile_options(${test_target} PRIVATE $<$<CONFIG:Debug>:/MDd> $<$<CONFIG:Release>:/MD>)
target_compile_options(${test_target_catch2} PRIVATE $<$<CONFIG:Debug>:/MDd> $<$<CONFIG:Release>:/MD>)
work?

**Comment by Alex-PLACET** [comment]
I think we should try to pass to -DMSVC_RUNTIME_LIBRARY=MultiThreadedDLL when on Windows we use Conda.
It means that we can remove the latest modification about to set MD and _ITERATOR_DEBUG_LEVEL, and in the windows ci matrix: https://github.com/man-group/sparrow/blob/main/.github/workflows/windows.yml#L116
replace
`echo "SPARROW_ADDITIONAL_OPTIONS=-DSPARROW_TARGET_32BIT=ON" >> $GITHUB_ENV`
by
`echo "SPARROW_ADDITIONAL_OPTIONS=-DSPARROW_TARGET_32BIT=ON -DMSVC_RUNTIME_LIBRARY=MultiThreadedDLL" >> $GITHUB_ENV`
I'm not sure it's sufficient to handle _ITERATOR_DEBUG_LEVEL
On my side I'm close to use Ubuntu 22 and a recent version of catch2 on exotic architectures

File: .github/workflows/windows.yml [hunk 2]
```diff
@@ -118,14 +118,15 @@ jobs:
       run: |
         echo "SPARROW_DEPS_PREFIX=$CONDA_PREFIX" >> $GITHUB_ENV
         echo "SPARROW_INSTALL_PREFIX=$CONDA_PREFIX" >> $GITHUB_ENV
-        echo "SPARROW_ADDITIONAL_OPTIONS=-DSPARROW_TARGET_32BIT=OFF" >> $GITHUB_ENV
+        echo "SPARROW_ADDITIONAL_OPTIONS=-DSPARROW_TARGET_32BIT=ON -DMSVC_RUNTIME_LIBRARY=MultiThreadedDLL" >> $GITHUB_ENV
```
Comment (Alex-PLACET) [inline .github/workflows/windows.yml:121]:
```suggestion
        echo "SPARROW_ADDITIONAL_OPTIONS=-DSPARROW_TARGET_32BIT=OFF -DMSVC_RUNTIME_LIBRARY=MultiThreadedDLL" >> $GITHUB_ENV
```

File: .github/workflows/windows.yml [hunk 3]
```diff
@@ -118,14 +118,15 @@ jobs:
...
 
     - name: Set dependencies install prefix dir for 32bit
       if: matrix.target-arch.name == 'Win32'
       run: |
         echo "SPARROW_DEPS_PREFIX=$VCPKG_INSTALLED_DIR" >> $GITHUB_ENV
         echo "SPARROW_INSTALL_PREFIX=$VCPKG_INSTALLED_DIR" >> $GITHUB_ENV
         echo "VCPKG_TARGET_TRIPLET='x86-windows'" >> $GITHUB_ENV
+        echo "VCPKG_BUILD_TYPE='Release'" >> $GITHUB_ENV
```
Comment (Alex-PLACET) [inline .github/workflows/windows.yml:129]: To remove

**Comment by Alex-PLACET** [comment]
To avoid the issues of Catch2 version on exotics architecture, can you introduce this following mechanism:
``` cmake
set(DOWNLOAD_TEST_DEPENDENCIES ON CACHE BOOL "Download test dependencies")
if(DOWNLOAD_TEST_DEPENDENCIES)
    Include(FetchContent)
    message(STATUS "Downloading doctest")
    FetchContent_Declare(doctest GIT_REPOSITORY https://github.com/onqtam/doctest       GIT_TAG v2.4.11)
    message(STATUS "Downloading catch2")
    FetchContent_Declare(catch2  GIT_REPOSITORY https://github.com/catchorg/Catch2.git  GIT_TAG v3.7.1)
    FetchContent_MakeAvailable(doctest catch2)
else()
    find_package(doctest 2 REQUIRED)
    find_package(catch2 3 REQUIRED)
endif()

```

**Comment by Gulin7** [comment]
Yes, I'll add both changes on Monday.

**Comment by Alex-PLACET** [comment]
@Gulin7 After some internal discussion, let's drop the CMake fetch solution.
Can you modify the the exotics arch CI to pull, compile and install the right Catch2 version ?

**Comment by Alex-PLACET** [comment]
@Gulin7 Can you add me as collaborator on your branch ?

**Comment by Alex-PLACET** [comment]
The CI is green 👍

File: CMakeLists.txt [hunk 4]
```diff
@@ -86,6 +86,8 @@ OPTION(BUILD_TESTS "Build sparrow test suite" OFF)
 OPTION(BUILD_DOCS  "Build sparrow documentation" OFF)
 OPTION(BUILD_EXAMPLES "Build sparrow examples" OFF)
 OPTION(USE_DATE_POLYFILL "Use date polyfill implementation" ON)
+set(DEPENDENCIES_CMAKE_FOLDER "External dependencies")
+OPTION(FETCH_DEPENDENCIES_WITH_CMAKE "Fetch dependencies with CMake: Can be OFF, ON, or MISSING, in this case CMake download only dependencies which are not previously found." OFF)
```
[the remark this answers is not in the document]
Comment (Alex-PLACET) [reply CMakeLists.txt]: Ok done

**Comment by Alex-PLACET** [comment]
@Gulin7 Thank you for your work!
The issue was more difficult than expected because of some build system and CI problem.

**Pull request merged**